File: Universal Presenter Remote/Universal Presenter Remote WatchKit Extension/DBZ_WKLoginView.m
//
//  DBZ_WKLoginView.m
//  Universal Presenter Remote
//
//  Created by Brendan Boyle on 3/15/15.
//  Copyright (c) 2015 DBZ Technology. All rights reserved.
//

#import "DBZ_WKLoginView.h"
#import "MMWormhole.h"
#import "UPR-Swift.h"

@interface DBZ_WKLoginView()

@end

@implementation DBZ_WKLoginView

NSTimer *checkInterfaceTimer;

bool firstTime = true;

- (void)awakeWithContext:(id)context {
    [super awakeWithContext:context];
    
    // Configure interface objects here.
}

- (void)willActivate {
    // This method is called when watch view controller is about to be visible to user
    [super willActivate];
    
    self.wormhole = [[MMWormhole alloc] initWithApplicationGroupIdentifier:@"group.com.dbztech.Universal-Presenter-Remote.wormhole" optionalDirectory:@"wormhole"];
    
    NSDictionary *requst = @{@"request":@"Startup"};
    
    /*
    [DBZ_WKLoginView openParentApplication:requst reply:^(NSDictionary *replyInfo, NSError *error) {
        
        if (error) {
            NSLog(@"%@", error);
        } else {
            
            NSLog(@"%@",[replyInfo objectForKey:@"response"]);
        }
        
    }];
     */
    
    if (firstTime) {
        requst = @{@"request":@"Refresh"};
        
        /*
        [DBZ_WKLoginView openParentApplication:requst reply:^(NSDictionary *replyInfo, NSError *error) {
            
            if (error) {
                NSLog(@"%@", error);
            } else {
                
                NSLog(@"%@",[replyInfo objectForKey:@"response"]);
            }
            
        }];
        firstTime = false;
         */
    }
    
    
    
    // Become a listener for changes to the wormhole for the button message
    [self.wormhole listenForMessageWithIdentifier:@"UPRWatchData" listener:^(id messageObject) {
        // The number is identified with the buttonNumber key in the message object
        NSNumber *number = [messageObject valueForKey:@"token"];
        NSNumber *enabled = [messageObject valueForKey:@"connectEnabled"];
        NSString *title = [messageObject valueForKey:@"connectTitle"];
        if (![[number stringValue] isEqualToString:@"0"]) {
            if ([[number stringValue]  isEqual: @""]) {
                NSDictionary *requst = @{@"request":@"Refresh"};
                
                /*
                [DBZ_WKLoginView openParentApplication:requst reply:^(NSDictionary *replyInfo, NSError *error) {
                    
                    if (error) {
                        NSLog(@"%@", error);
                    } else {
                        
                        NSLog(@"%@",[replyInfo objectForKey:@"response"]);
                    }
                    
                }];
                 */
            }
            _tokenLabel.text = [number stringValue];
        } else {
            _tokenLabel.text = @"...";
        }
        
        [_connectButton setTitle:title];
        if ([enabled isEqualToNumber:@(1)]) {
            [_connectButton setEnabled:YES];
        } else {
            [_connectButton setEnabled:NO];
        }
    }];
    
    //checkInterfaceTimer = [NSTimer scheduledTimerWithTimeInterval:2.0 target:self selector:@selector(checkInterface:) userInfo:nil repeats:YES];
}

- (void)didDeactivate {
    // This method is called when watch view controller is no longer visible
    [super didDeactivate];
}

-(void)checkInterface:(NSTimer*)timer {
    self.wormhole = [[MMWormhole alloc] initWithApplicationGroupIdentifier:@"group.com.dbztech.Universal-Presenter-Remote.wormhole" optionalDirectory:@"wormhole"];
    
    NSDictionary *requst = @{@"request":@"Startup"};
    
    /*
    [DBZ_WKLoginView openParentApplication:requst reply:^(NSDictionary *replyInfo, NSError *error) {
        
        if (error) {
            NSLog(@"%@", error);
        } else {
            
            NSLog(@"%@",[replyInfo objectForKey:@"response"]);
        }
        
    }];
    */
}

- (IBAction)refreshToken {
    NSDictionary *requst = @{@"request":@"Refresh"};
    
    /*
    [DBZ_WKLoginView openParentApplication:requst reply:^(NSDictionary *replyInfo, NSError *error) {
        
        if (error) {
            NSLog(@"%@", error);
        } else {
            
            NSLog(@"%@",[replyInfo objectForKey:@"response"]);
        }
        
    }];
     */
}

- (IBAction)openInstructions {
    NSLog(@"Open Instructions");
    [self presentControllerWithName:@"Instructions" context: nil];
}
@end

File: Universal Presenter Remote/Universal Presenter Remote WatchKit Extension/DBZ_WKLoginView.swift
//
//  DBZ_WKLoginController.swift
//  Universal Presenter Remote
//
//  Created by Brendan Boyle on 4/25/16.
//  Copyright © 2016 DBZ Technology. All rights reserved.
//

import WatchKit
import Foundation


class DBZ_WKLoginView: WKInterfaceController {

    @IBOutlet var tokenLabel: WKInterfaceLabel!
    @IBOutlet var connectButton: WKInterfaceButton!
    
    var refreshTimer: Timer?
    var needsSetup: Bool = true

    override func willActivate() {
        // This method is called when watch view controller is about to be visible to user
        super.willActivate()
        
        connectButton.setEnabled(false)
        
        if (!DBZ_ServerCommunication.setup()) {
            DBZ_ServerCommunication.setupUid()
        }
        
        if needsSetup {
            NotificationCenter.default.addObserver(self, selector: #selector(refreshInterface), name: NSNotification.Name(rawValue: "Refresh"), object: nil)
            NotificationCenter.default.addObserver(self, selector: #selector(updateInterface), name: NSNotification.Name(rawValue: "UpdateInterface"), object: nil)
            needsSetup = false
        }
        
        DBZ_ServerCommunication.checkToken()
        ensureTimer()
    }
    
    func ensureTimer() {
        if refreshTimer == nil || !refreshTimer!.isValid {
            refreshTimer = Timer.scheduledTimer(timeInterval: 2.5, target: self, selector: #selector(refreshLocal), userInfo: nil, repeats: true)
        }
    }
    
    @objc func updateInterface(_ notification:Notification) {
        updateTokenLabel()
        switch (DBZ_ServerCommunication.controlmode()) {
            case 0:
                //Button Connecting... NO
                connectButton.setEnabled(false)
                connectButton.setTitle("Connecting...")
                connectButton.setBackgroundColor(nil)
                break
            case 1:
                //Button Waiting... NO
                connectButton.setEnabled(false)
                connectButton.setTitle("Waiting...")
                connectButton.setBackgroundColor(nil)
                break
            case 2:
                //Button Begin YES
                connectButton.setEnabled(true)
                connectButton.setTitle("Begin")
                connectButton.setBackgroundColor(UIColor(named: "Primary"))
                DBZ_ServerCommunication.startSession()
                refreshTimer?.invalidate()
                break
            default:
                break
        }
    }
    
    func updateTokenLabel() {
        let token = DBZ_ServerCommunication.temptoken()
        if (token > 10) {
            //Set token label
            tokenLabel.setText(String(token))
        } else {
            tokenLabel.setText("...")
        }
    }
    
    @objc func refreshInterface(_ notification:Notification) {
        DBZ_ServerCommunication.checkToken()
    }
    
    @objc func refreshLocal() {
        DBZ_ServerCommunication.checkToken()
        self.updateTokenLabel()
    }

    @IBAction func reloadButton() {
        DBZ_ServerCommunication.setupUid()
        updateTokenLabel()
        connectButton.setEnabled(false)
        DBZ_ServerCommunication.checkToken()
        ensureTimer()
    }
}
